src/html/classic/ng/src/web/pages/extras/trashcanpage.js

```diff
@@ -0,0 +1,493 @@
+import ScanConfigsTable from '../scanconfigs/table.js';
+import CredentialsTable from '../credentials/table.js';
+import FiltersTable from '../filters/table.js';
+import GroupsTable from '../groups/table.js';
+import NotesTable from '../notes/table.js';
+import OverridesTable from '../overrides/table.js';
+import PermissionsTable from '../permissions/table.js';
+import PortListsTable from '../portlists/table.js';
+import ReportFormatsTable from '../reportformats/table.js';
+import RolesTable from '../roles/table.js';
+import ScannersTable from '../scanners/table.js';
+import SchedulesTable from '../schedules/table.js';
+import TagsTable from '../tags/table.js';
+import TargetsTable from '../targets/table.js';
+import TasksTable from '../tasks/table.js';
+
+import Table from '../../components/table/strippedtable.js';
+import TableBody from '../../components/table/body.js';
+import TableData from '../../components/table/data.js';
+import TableRow from '../../components/table/row.js';
+import TableHead from '../../components/table/head.js';
+import TableHeader from '../../components/table/header.js';
+import {Col} from 'glamorous';
+
+import CollectionList from 'gmp/collection/collectionlist.js';
+
+import Layout from '../../components/layout/layout.js';
+import Section from '../../components/section/section.js';
+
+import InnerLink from '../../components/link/innerlink.js';
+import LinkTarget from '../../components/link/target.js';
+
+import Button from '../../components/form/button.js';
+import HelpIcon from '../../components/icon/helpicon.js';
+
+import Loading from '../../components/loading/loading.js';
+
+import {map, is_defined} from 'gmp/utils.js';
+import PropTypes from '../../utils/proptypes.js';
+import withContext from '../../utils/withContext.js';
+
+
+const ToolBarIcons = () => (
+  <HelpIcon
+    page="trashcan"
+    title={_('Help: Trashcan')}
+  />
+);
+
+const EmptyTrashButton = ({onClick}) => (
+  <Layout align="right">
+    <Button
+      onClick={onClick}
+    >{_('Empty Trash')}</Button>
+  </Layout>
+);
+
+class Trashcan extends React.Component {
+
+  constructor(...args) {
+    super(...args);
+    this.state = {
+      trash_data: undefined,
+      contents_table: [],
+    };
+
+    this.createContentRow = this.createContentRow.bind(this);
+    this.createContentsTable = this.createContentsTable.bind(this);
+    this.getTrash = this.getTrash.bind(this);
+    this.handleEmptyTrashcan = this.handleEmptyTrashcan.bind(this);
+    this.handleDelete = this.handleDelete.bind(this);
+    this.handleRestore = this.handleRestore.bind(this);
+  }
+
+  componentDidMount() {
+      this.getTrash();
+  }
+
+  getTrash() {
+    const {gmp} = this.props;
+    const data = Promise.all([
+      gmp.trashcan.readTrashcan().then(response => {
+        const trash_data = response.data;
+        const contents_table = this.createContentsTable(trash_data);
+        this.setState({trash_data, contents_table});
```

Comment:
Please don't put a react element into the state. The react element should only be created in the render method.

Reply:
createContentsTable() is now a function which gets called from the render() method, without the detour through the state.